File: mobile/ios/src/data/session-manager.swift
import Foundation
import Hydra
import AuthenticationServices
import FBSDKLoginKit

private struct SessionV1: Codable {
  var token: String?
  var refreshToken: String?
  var lastVenue: String?
  var cachedVenueLocations: [String:String]?
}

private struct SessionV2: Codable {
  var pushToken: String?
  var hasSkippedOnboarding: Bool?
}

private struct SessionV3: Codable {
  var phoneNumber: String?
}

private struct SessionV4: Codable {
  var authTokenExpiry: Date?
  var refreshTokenExpiry: Date?
}

private struct SessionV5: Codable {
  var source: SessionSource?
}

private struct SessionImpl: Session {
  var version: Semver
  var v1: SessionV1?
  var v2: SessionV2?
  var v3: SessionV3?
  var v4: SessionV4?
  var v5: SessionV5?

  var lastVenue: String? {
    get { return v1?.lastVenue }
    set { v1?.lastVenue = newValue }
  }
  var cachedVenueLocations: [String : String]? {
    get { return v1?.cachedVenueLocations }
    set { v1?.cachedVenueLocations = newValue }
  }
  var authenticationToken: String? {
    get { return v1?.token }
    set { v1?.token = newValue }
  }
  var refreshToken: String? {
    get { return v1?.refreshToken }
    set { v1?.refreshToken = newValue }
  }
  var pushToken: String? {
    get { return v2?.pushToken }
    set { v2?.pushToken = newValue }
  }
  var hasSkippedOnboarding: Bool? {
    get { return v2?.hasSkippedOnboarding }
    set { v2?.hasSkippedOnboarding = newValue }
  }
  var phoneNumber: String? {
    get { return v3?.phoneNumber }
    set { v3?.phoneNumber = newValue }
  }
  var authTokenExpiry: Date? {
    get { return v4?.authTokenExpiry }
    set { v4?.authTokenExpiry = newValue }
  }
  var refreshTokenExpiry: Date? {
    get { return v4?.refreshTokenExpiry }
    set { v4?.refreshTokenExpiry = newValue }
  }
  var source: SessionSource? {
    get { return v5?.source }
    set { v5?.source = newValue }
  }
}

protocol SessionProvider {
  func initialise() -> Promise<Void>
  var isExpired: Bool {get}
  var isAuthenticated: Bool {get}
  var authToken: String? {get}
  var accountDescription: String? {get}
  var firstName: String? {get}
  var lastName: String? {get}
  func registerBespokeSession(_ session: Any?)
}

protocol SessionManager {
  var isAuthenticated: Bool { get }
  var isExpired: Bool { get }
  var authToken: String? { get }
  var accountDescription: String? { get }
  var firstName: String? { get }
  var lastName: String? { get }
  var authType: String? { get }
  func registerBespokeSession(_ bespokeSession: Any?)
  func restore () -> Promise<Session>
  func persist (session: Session) -> Promise<Void>
  func purge () -> Promise<Void>
  var currentSession: Session? { get }
  var currentProfile: UserProfile? { get }
  func updateProfile(_ userProfile: UserProfile)
}

class SessionManagerImpl: SessionManager {
  fileprivate static let LATEST_VERSION = Semver(version: "0.0.5")
  fileprivate(set) var currentSession: Session?
  fileprivate let facebookSessionProvider: SessionProvider
  fileprivate let appleSessionProvider: SessionProvider
  fileprivate(set) var currentProfile: UserProfile?

  init(facebookSessionProvider: SessionProvider, appleSessionProvider: SessionProvider) {
    self.facebookSessionProvider = facebookSessionProvider
    self.appleSessionProvider = appleSessionProvider
  }

  var isAuthenticated: Bool {
    get {
      guard let session = currentSession else {
        return false
      }

      guard let source = session.source else {
        return false
      }

      switch source {
      case .apple:
        return appleSessionProvider.isAuthenticated
      case .facebook:
        return facebookSessionProvider.isAuthenticated
      }
    }
  }

  var isExpired: Bool {
    get {
      guard let session = currentSession else {
        return false
      }

      guard let source = session.source else {
        return false
      }

      switch source {
      case .apple:
        return appleSessionProvider.isExpired
      case .facebook:
        return facebookSessionProvider.isExpired
      }
    }
  }

  var authToken: String? {
    get {
      guard let session = currentSession else {
        return nil
      }

      guard let source = session.source else {
        return nil
      }

      switch source {
      case .apple:
        return appleSessionProvider.authToken
      case .facebook:
        return facebookSessionProvider.authToken
      }
    }
  }

  var authType: String? {
    get {
      guard let session = currentSession else {
        return nil
      }

      guard let source = session.source else {
        return nil
      }

      switch source {
      case .apple:
        return "apple"
      case .facebook:
        return "facebook"
      }
    }
  }

  var accountDescription: String? {
    get {
      guard let session = currentSession else {
        return nil
      }

      guard let source = session.source else {
        return nil
      }

      switch source {
      case .apple:
        return appleSessionProvider.accountDescription
      case .facebook:
        return facebookSessionProvider.accountDescription
      }
    }
  }

  var firstName: String? {
    get {
      guard let session = currentSession else {
        return nil
      }

      guard let source = session.source else {
        return nil
      }

      switch source {
      case .apple:
        return appleSessionProvider.firstName ?? currentProfile?.firstName
      case .facebook:
        return facebookSessionProvider.firstName ?? currentProfile?.firstName
      }
    }
  }

  var lastName: String? {
    get {
      guard let session = currentSession else {
        return nil
      }

      guard let source = session.source else {
        return nil
      }

      switch source {
      case .apple:
        return appleSessionProvider.lastName ?? currentProfile?.lastName
      case .facebook:
        return facebookSessionProvider.lastName ?? currentProfile?.lastName
      }
    }
  }

  func registerBespokeSession(_ bespokeSession: Any?) {
    guard let session = currentSession else {
      return
    }

    guard let source = session.source else {
      return
    }

    switch source {
    case .apple:
      appleSessionProvider.registerBespokeSession(bespokeSession)
    case .facebook:
      facebookSessionProvider.registerBespokeSession(bespokeSession)
    }
  }

  func restore () -> Promise<Session> {
    return async({ _ -> Session in
      try await (self.facebookSessionProvider.initialise())
      try await (self.appleSessionProvider.initialise())

      let decoder = JSONDecoder.defaultDecoder()
      let documentsUrl = try FileManager.default.url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
      let sessionUrl = documentsUrl.appendingPathComponent("session.json")

      if !FileManager.default.fileExists(atUrl: sessionUrl) {
        let session = SessionImpl(version: SessionManagerImpl.LATEST_VERSION, v1: SessionV1(), v2: SessionV2(), v3: SessionV3(), v4: SessionV4(), v5: SessionV5())
        try await(self.persist(session: session))

        return session
      } else {
        guard let data = try? Data(contentsOf: sessionUrl) else {
          let session = SessionImpl(version: SessionManagerImpl.LATEST_VERSION, v1: SessionV1(), v2: SessionV2(), v3: SessionV3(), v4: SessionV4(), v5: SessionV5())
          try await(self.persist(session: session))

          return session
        }

        guard var session = try? decoder.decode(SessionImpl.self, from: data) else {
          let session = SessionImpl(version: SessionManagerImpl.LATEST_VERSION, v1: SessionV1(), v2: SessionV2(), v3: SessionV3(), v4: SessionV4(), v5: SessionV5())
          try await(self.persist(session: session))

          return session
        }

        if session.v1 == nil {
          session.v1 = SessionV1()
        }

        if session.v2 == nil {
          session.v2 = SessionV2()
        }

        if session.v3 == nil {
          session.v3 = SessionV3()
        }

        if session.v4 == nil {
          session.v4 = SessionV4()
        }

        if session.v5 == nil {
          session.v5 = SessionV5()
        }

        return session
      }
    }).then(in: .main, { [weak self] session -> Session in
      self?.currentSession = session
      return session
    })
  }

  func persist (session: Session) -> Promise<Void> {
    return async({ _ -> Void in
      guard let session = session as? SessionImpl else {
        throw ApplicationError(message: "Provided session was not recognised, you should not subclass the Session protocol.")
      }

      let encoder = JSONEncoder.defaultEncoder()
      let documentsUrl = try FileManager.default.url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
      let sessionUrl = documentsUrl.appendingPathComponent("session.json")

      try encoder.encode(self.upgrade(sessionData: session)).write(to: sessionUrl)
    }).then(in: .main, { [weak self] in
      self?.currentSession = session
    })
  }

  func purge () -> Promise<Void> {
    currentSession = SessionImpl(version: SessionManagerImpl.LATEST_VERSION, v1: SessionV1(), v2: SessionV2(), v3: SessionV3(), v4: SessionV4(), v5: SessionV5())
    currentProfile = nil
    return async({ _ -> Void in
      let documentsUrl = try FileManager.default.url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
      let sessionUrl = documentsUrl.appendingPathComponent("session.json")

      guard FileManager.default.fileExists(atUrl: sessionUrl) else { return }
      try FileManager.default.removeItem(at: sessionUrl)
    }).catch { _ in }.then(in: .main, {
      LoginManager().logOut()
    })
  }

  fileprivate func upgrade(sessionData session: SessionImpl) -> SessionImpl {
    // TODO: When moving to a v2, add a method in here to perform data migration.
    return session
  }

  func updateProfile(_ userProfile: UserProfile) {
    currentProfile = userProfile
  }
}

File: mobile/android/app/src/main/java/io/meshtech/koasta/data/internal/SessionManager.kt
package io.meshtech.koasta.data.internal

import com.facebook.login.LoginManager
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.squareup.moshi.FromJson
import com.squareup.moshi.JsonClass
import com.squareup.moshi.Moshi
import com.squareup.moshi.ToJson
import com.squareup.moshi.adapters.Rfc3339DateJsonAdapter
import io.meshtech.koasta.BuildConfig
import io.meshtech.koasta.GlobalApplication
import io.meshtech.koasta.data.ISessionManager
import io.meshtech.koasta.data.Session
import io.meshtech.koasta.net.model.UserProfile
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import net.swiftzer.semver.SemVer
import java.io.File
import java.util.*

internal class SemVerAdapter {
  @ToJson
  fun toJson(semver: SemVer): String {
    return semver.toString()
  }

  @FromJson
  fun fromJson(semver: String): SemVer {
    return SemVer.parse(semver)
  }
}

@JsonClass(generateAdapter = true)
internal data class SessionV1 constructor(
  var token: String? = null,
  var refreshToken: String? = null,
  var lastVenue: String? = null,
  var cachedVenueLocations: Map<String, String>? = null,
  var pushToken: String? = null,
  var hasSkippedOnboarding: Boolean? = null,
  var phoneNumber: String? = null,
  var authTokenExpiry: Date? = null,
  var refreshTokenExpiry: Date? = null
)

@JsonClass(generateAdapter = true)
internal data class SessionV2 constructor(
  var source: Int? = null
)

@JsonClass(generateAdapter = true)
internal data class SessionHolder (
  var version: SemVer? = null,
  var v1: SessionV1? = null,
  var v2: SessionV2? = null
)

internal class SessionImpl(holder: SessionHolder? = null) : Session {
  var version: SemVer? = holder?.version
  var v1: SessionV1? = holder?.v1
  var v2: SessionV2? = holder?.v2

  fun toHolder(): SessionHolder {
    return SessionHolder(version, v1, v2)
  }

  override var authenticationToken: String?
    get() = v1?.token
    set(value) {
      v1?.token = value
    }

  override var refreshToken: String?
    get() = v1?.refreshToken
    set(value) {
      v1?.refreshToken = value
    }

  override var lastVenue: String?
    get() = v1?.lastVenue
    set(value) {
      v1?.lastVenue = value
    }

  override var cachedVenueLocations: Map<String, String>?
    get() = v1?.cachedVenueLocations
    set(value) {
      v1?.cachedVenueLocations = value
    }

  override var pushToken: String?
    get() = v1?.pushToken
    set(value) {
      v1?.pushToken = value
    }

  override var hasSkippedOnboarding: Boolean?
    get() = v1?.hasSkippedOnboarding
    set(value) {
      v1?.hasSkippedOnboarding = value
    }

  override var phoneNumber: String?
    get() = v1?.phoneNumber
    set(value) {
      v1?.phoneNumber = value
    }

  override var authTokenExpiry: Date?
    get() = v1?.authTokenExpiry
    set(value) {
      v1?.authTokenExpiry = value
    }

  override var refreshTokenExpiry: Date?
    get() = v1?.refreshTokenExpiry
    set(value) {
      v1?.refreshTokenExpiry = value
    }

  override var source: Int?
    get() = v2?.source
    set(value) {
      v2?.source = value
    }

  fun copy(): SessionImpl {
    val ret = SessionImpl()

    ret.version = version?.copy()
    ret.v1 = v1?.copy()
    ret.v2 = v2?.copy()

    return ret
  }

  init {
    version = version ?: SessionManager.LATEST_VERSION
    v1 = v1 ?: SessionV1(null, null, null,
      null, null, null,
      null, null, null)
    v2 = v2 ?: SessionV2(
      null
    )
  }
}

class SessionManager(private val facebookProvider: IFacebookSessionProvider, private val googleProvider: IGoogleSessionProvider): ISessionManager {
  companion object {
    val LATEST_VERSION = SemVer(0, 0, 2)
  }

  private var session: SessionImpl = SessionImpl()
  private val moshi = Moshi.Builder()
    .add(Date::class.java, Rfc3339DateJsonAdapter().nullSafe())
    .add(SemVerAdapter())
    .build()
  override var currentProfile: UserProfile? = null

  override val currentSession: Session
    get() = session.copy()

  override val isAuthenticated: Boolean
    get() {
      if (session.source == null) {
        return false
      }

      return when (session.source) {
        1 -> {
          facebookProvider.isAuthenticated
        }
        3 -> {
          googleProvider.isAuthenticated
        }
        else -> {
          false
        }
      }
    }
  override val isExpired: Boolean
    get() {
      if (session.source == null) {
        return false
      }

      return when (session.source) {
        1 -> {
          facebookProvider.isExpired
        }
        3 -> {
          googleProvider.isExpired
        }
        else -> {
          false
        }
      }
    }
  override val authToken: String?
    get() {
      if (session.source == null) {
        return null
      }

      return when (session.source) {
        1 -> {
          facebookProvider.authToken
        }
        3 -> {
          googleProvider.authToken
        }
        else -> {
          null
        }
      }
    }
  override val authType: String?
    get() {
      if (session.source == null) {
        return null
      }

      return when (session.source) {
        1 -> {
          "facebook"
        }
        3 -> {
          "google"
        }
        else -> {
          null
        }
      }
    }

  override val accountDescription: String?
    get() {
      if (session.source == null) {
        return null
      }

      return when (session.source) {
        1 -> {
          facebookProvider.accountDescription
        }
        3 -> {
          googleProvider.accountDescription
        }
        else -> {
          null
        }
      }
    }

  override val firstName: String?
    get() {
      if (session.source == null) {
        return null
      }

      return when (session.source) {
        1 -> {
          facebookProvider.firstName ?: currentProfile?.firstName
        }
        3 -> {
          googleProvider.firstName ?: currentProfile?.firstName
        }
        else -> {
          null
        }
      }
    }

  override val lastName: String?
    get() {
      if (session.source == null) {
        return null
      }

      return when (session.source) {
        1 -> {
          facebookProvider.lastName ?: currentProfile?.lastName
        }
        3 -> {
          googleProvider.lastName ?: currentProfile?.lastName
        }
        else -> {
          null
        }
      }
    }

  fun registerBespokeSession(newSession: Any?) {
    if (session.source == null) {
      return
    }

    when (session.source) {
      1 -> {
        facebookProvider.registerBespokeSession(newSession)
      }
      3 -> {
        googleProvider.registerBespokeSession(newSession)
      }
    }
  }

  override suspend fun restore(): Session {
    return withContext(Dispatchers.IO) {
      val file = File(GlobalApplication.shared.filesDir, "session.json")
      if (file.exists()) {
        session = SessionImpl(moshi.adapter(SessionHolder::class.java).fromJson(file.readText()) ?: SessionHolder())
      }

      withContext(Dispatchers.Main) {
        this@SessionManager.session
      }
    }
  }

  override suspend fun persist(session: Session) {
    return withContext(Dispatchers.IO) {
      val file = File(GlobalApplication.shared.filesDir, "session.json")
      file.writeText(moshi.adapter(SessionHolder::class.java).toJson((session as SessionImpl).toHolder()))

      this@SessionManager.session = session
    }
  }

  override fun purge() {
    val file = File(GlobalApplication.shared.filesDir, "session.json")
    if (file.exists()) {
      file.delete()
    }
    session = SessionImpl(SessionHolder())
    LoginManager.getInstance().logOut()
    val signInClient = GoogleSignIn.getClient(GlobalApplication.shared, GoogleSignInOptions.Builder(
      GoogleSignInOptions.DEFAULT_SIGN_IN)
      .requestIdToken(BuildConfig.GOOGLE_API_KEY)
      .requestEmail()
      .build())
    signInClient.signOut()
    currentProfile = null
  }
}
